[YosieSYX] iP

YosieSYX opened the pull request:
# Cherry

> "Your mind is for having ideas, not holding them." – David Allen

Cherry frees your mind of having to remember things you need to do. It's:

- **Text-based**
- **Easy to learn**
- **SUPER FAST.**

## Getting Started

1. [Download it from here](https://github.com/YosieSYX/ip).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉

**And the best part? It's FREE!**

## Features

- [ ] Managing tasks
- [ ] Managing deadlines 
- [ ] Reminders *(coming soon)*

## For Java Programmers

If you're a Java programmer, you can use Cherry to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        new Cherry("./data/Tasks.txt").run();
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

YosieSYX committed:
Implemented level-0

YosieSYX committed:
Implemented level-0

YosieSYX committed:
Implemented level-0

YosieSYX committed:
Implemented level-0

YosieSYX committed:
Implemented level-1

YosieSYX committed:
Implemented level-2

YosieSYX committed:
Implemented Level-3

YosieSYX committed:
Implemented Level-4

YosieSYX committed:
Finished A-TextUiTesting

YosieSYX committed:
Implemeneted Level-5

YosieSYX committed:
Implemeneted Level-6

YosieSYX committed:
Updated Level-4

YosieSYX committed:
Used enum for weekday

YosieSYX committed:
Used enum for weekday

YosieSYX committed:
Used enum for weekday

YosieSYX committed:
Implement Level-7

YosieSYX committed:
Merge branch 'branch-Level-7'
merge 2 branches

YosieSYX committed:
Implement Level-8

YosieSYX committed:
partial Level-8

YosieSYX committed:
changed level-7

YosieSYX committed:
Merge branch 'branch-Level-7'
change this level implementation

YosieSYX committed:
Merge branch 'branch-Level-8'

YosieSYX committed:
fixed some error

YosieSYX committed:
implement Level-8

YosieSYX committed:
change this branch

YosieSYX committed:
Merge branch 'branch-Level-8'
finish Level-8

YosieSYX committed:
implement A-MoreOOP

YosieSYX committed:
Merge branch 'A-MoreOOP'
implement A-MoreOOP

YosieSYX committed:
implement A-Packages

YosieSYX committed:
Merge remote-tracking branch 'origin/add-gradle-support'
merge branch add-gradle-support

YosieSYX committed:
implement A-Gradle

YosieSYX committed:
implement A-JUnit

YosieSYX committed:
implement A-Jar

YosieSYX committed:
implement A-JavaDoc

YosieSYX committed:
implement A-CodingStandard

YosieSYX committed:
implement Level-9

YosieSYX committed:
Merge branch 'branch-A-JavaDoc'
merge branch 'branch-A-JavaDoc'

YosieSYX committed:
resolve merge conflicts

YosieSYX committed:
merge Level-9

deseansoh commented on src/main/java/cherry/main/Cherry.java:
```diff
@@ -0,0 +1,125 @@
+                    ui.showReceivedMessage("Now you have " + count + " tasks in the list!");
+                    storage.save(tasks.toList());
+                } else {
+                    ui.showErrorMessage("please give a valid todo");
```
Perhaps a switch statement could be utilised here to keep the code cleaner

deseansoh commented on src/main/java/cherry/main/Deadline.java:
```diff
@@ -0,0 +1,90 @@
+ */
+public class Deadline extends Task {
+
+    protected Object by;
```
```suggestion
    protected Object byDeadline;
```
I think that changing the variable name here could make it clearer for understanding

deseansoh commented on src/main/java/cherry/main/TaskList.java:
The naming of your methods are all very consistent and easy to understand!

deseansoh commented on src/main/java/cherry/utils/Ui.java:
```diff
@@ -0,0 +1,66 @@
+     * @param message The message to be displayed.
+     */
+    public void showReceivedMessage(String message) {
+        System.out.println(message);
```
Nice use of methods to keep the rest of the codebase clean

deseansoh reviewed (commented):
The codebase is clean and I had a good time reviewing it!

YosieSYX committed:
implement A-CheckStyle and fix checkstyle error

ho-thanh-und commented:
The codebase is well structured and was intuitive to understand! I had fun reviewing this PR and learning from you!

ho-thanh-und commented on src/main/java/cherry/main/Cherry.java:
```diff
@@ -0,0 +1,142 @@
+            } else if (input.startsWith("find")) {
+                String[] parts = parser.parseFind(input);
+                tasks.findTasks(parts[1]);
+            } else if (input.startsWith("by")) {
+                String OriDate = input.split("by")[1].trim();
+                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM dd yyyy");
```
I realized for the 'OriDate' variable, it should follow camelCase as its a String variable, so maybe oriDate?

ho-thanh-und commented on src/main/java/cherry/main/Deadline.java:
```diff
@@ -0,0 +1,90 @@
+        return this.by;
+    }
+
+    /**
+     * Parses the input string to determine if it represents a date or a day of the week.
+     * If the input is a valid day of the week, it returns the corresponding Day enum.
+     * If the input is a valid date in the format yyyy-MM-dd, it returns a LocalDate object.
+     * If the input is neither, it returns null.
+     *
+     * @param input The input string to parse.
+     * @return A LocalDate object, a Day enum, or null if the input is invalid.
+     */
+    private Object parseDateOrDay(String input) {
```
I like the detailed comments you provided. This made it easier to understand your code!

ho-thanh-und commented on src/main/java/cherry/utils/Day.java:
```diff
@@ -0,0 +1,8 @@
+/**
+ * The Day enum represents the days of the week.
+ */
+public enum Day {
+    Monday, Tuesday, Wednesday, Thursday, Friday, Saturday, Sunday
+}
```
I think this is convenient! You reminded me of implementing Enum which I haven't done!

ho-thanh-und commented on src/main/java/cherry/utils/Storage.java:
```diff
@@ -0,0 +1,77 @@
+        File file = new File(filePath);
+        File parentDir = file.getParentFile();
+
+        try {
+            if (parentDir != null && !parentDir.exists() && !parentDir.mkdirs()) {
+                System.err.println("Failed to create directory: " + parentDir.getAbsolutePath());
+            }
+
+            if (!file.exists()) {
+                if (file.createNewFile()) {
+                    System.out.println("New task file created: " + filePath);
+                } else {
+                    System.err.println("Failed to create file: " + filePath);
+                }
+                return tasks;
+            }
+        } catch (IOException e) {
+            System.err.println("Error while initializing storage");
+            return tasks;
+        }
```
I think this try-catch block can be more concise by separating the handling of directory correction and file creation!

YosieSYX committed:
implement Level-10

YosieSYX committed:
minor changes

YosieSYX committed:
implement A-Varargs

Billy026 commented on src/main/java/cherry/utils/Storage.java:
```diff
@@ -0,0 +1,77 @@
+            System.err.println("Error reading from file: " + filePath);
+        }
+        return tasks;
+    }
```
I think these try-catch blocks can be divided into two functions for a stronger SLAP.

Billy026 commented on src/main/java/cherry/main/Cherry.java:
```diff
@@ -0,0 +1,121 @@
+            return tasks.findTasks(parts);
+        } else if (input.startsWith("by")) {
+            try {
+                String OriDate = input.split("by")[1].trim();
+                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM dd yyyy");
+                LocalDate date = LocalDate.parse(OriDate, formatter);
+
+                StringBuilder result = new StringBuilder();
+                for (Task task : tasks.toList()) {
+                    if (task instanceof Deadline) {
+                        Deadline deadline = (Deadline) task;
+                        if (date.equals(deadline.getDateOrDay())) {
+                            result.append(deadline).append("\n");
+                        }
+                    } else if (task instanceof Events) {
+                        Events event = (Events) task;
+                        if (date.equals(event.getEndDate())) {
+                            result.append(event).append("\n");
+                        }
+                    }
+                }
+                return result.length() > 0 ? result.toString() : "No tasks found for that date.";
```
Perhaps you could extract this to another function for stronger SLAP?

Billy026 commented on src/main/java/cherry/main/Cherry.java:
```diff
@@ -0,0 +1,121 @@
+            tasks.addTask(new Events(parts[0], parts[1], parts[2]));
+            storage.save(tasks.toList());
+            return "Added: " + input + "\nNow you have " + tasks.count() + " tasks in the list!";
+        }
```
Perhaps you can have a function for each operation on tasks for a stronger SLAP?

Billy026 reviewed (commented):
Your code doesn't seem to have many problems with nesting problems or too-long methods (except getResponse but I can't think of any solution).  Perhaps you can extract code from certain functions into new functions to facilitate stronger SLAP?

YosieSYX committed:
add assertions

YosieSYX committed:
Improve code quality
Extract out methods to comply with Single Level Abstraction Principle
and deal with the unusual cases as early as possible.
This improves readability of code.

YosieSYX committed:
Improve code quality

Extract out functions to comply with the Single Level of Abstraction Principle and restructure code to handle unusual conditions first.

Doing this improves readability of the code.

YosieSYX committed:
Employ Java Streams

Use Java Streams when processing keyword String and finding the
corresponding tasks in task list.

Let's employ streams in code to simplify long nested loops.

YosieSYX committed:
Merge pull request #1 from YosieSYX/branch-A-Assertions

Add assertions

YosieSYX committed:
merge master to update with master

YosieSYX committed:
Merge pull request #2 from YosieSYX/branch-A-CodeQuality

Improve code quality

YosieSYX committed:
Merge pull request #3 from YosieSYX/branch-A-Streams

Employ Java Streams

YosieSYX committed:
set up continuous integration

YosieSYX committed:
Add tag feature

Enable users to add tags to tasks.

This helps users better organize and distinguish tasks.

YosieSYX committed:
improve GUI

improve GUI to make it more user friendly

YosieSYX committed:
add TaskListTest

YosieSYX committed:
Use AI

Use AI to generate JavaDoc comment.

YosieSYX committed:
Personalise chatbox

YosieSYX committed:
Display welcome message

YosieSYX committed:
Set chatbox name

YosieSYX committed:
Fix event handling

YosieSYX committed:
Add user guide

YosieSYX committed:
Update user guide

YosieSYX committed:
Fix user guide
